### DISCUSSION TITLE
Build from source on Windows

### DISCUSSION THREAD

**Discussion by kbengs** [pdfarranger/pdfarranger#991] [General]
The steps in Win32.md currently fails. One reason is that the cx_freeze that is used does not support python 3.11.
I made a commit https://github.com/kbengs/pdfarranger/commit/f78fa4c447c2e57f532948f777ab383cdc615cd7 which update the instructions and setup_win32.py

But this require that the docker file is updated too.  @jeromerobert what do you think about that, is that something you see as fun work? :smile:

**Comment by jeromerobert** [comment]
I wouldn't say it's a fun job, but it was already at the top of my to-do list. We're currently on pikepdf 6 and we really need to upgrade to 8. I start working on it right now.

Oh ! am I wrong or does your commit mean that pip install pikepdf finally works under mingw64 ? That will make life much easier.

**Comment by jeromerobert** [comment]
Well no `pip install pikepdf` is not that simple on mingw64.

```
  adding license file 'LICENSE.txt'
  writing manifest file 'src\pikepdf.egg-info\SOURCES.txt'
  copying src\pikepdf\_core.pyi -> build\lib.mingw_x86_64-cpython-311\pikepdf
  copying src\pikepdf\py.typed -> build\lib.mingw_x86_64-cpython-311\pikepdf
  running build_ext
  building 'pikepdf._core' extension
  error: --plat-name must be one of ('win32', 'win-amd64', 'win-arm32', 'win-arm64')
  [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pikepdf
```

**Comment by jeromerobert** [comment]
Work in progress but enough to create the .zip and .msi. Not tested yet. https://github.com/pdfarranger/pdfarranger-docker-ci/commit/375c618b3060038cbbf158df2e687a078ff7dd45

**Comment by kbengs** [comment]
> Well no pip install pikepdf is not that simple on mingw64.

I tested again and now it fails for me too. I guess I had something left from a old install that made it work for me earlier.
But this seems to work:
```
export PATH=$PATH:/c/msys64/mingw64/bin
SETUPTOOLS_USE_DISTUTILS=stdlib /mingw64/bin/python3 -m pip install --user pikepdf
```

This time I had problem with installing darkdetect though.

https://www.msys2.org/docs/python/#known-issues

**Comment by jeromerobert** [comment]
Ok I'll test it. In the meantime I pushed an temporary version:
https://github.com/pdfarranger/pdfarranger/actions/runs/6459683695

**Comment by jeromerobert** [comment]
It currently gives: `ModuleNotFoundError: No module named dateutil`, which should be easy to fix.

**Reply by jeromerobert** [reply]
Fixed

**Comment by jeromerobert** [comment]
pdfarranger now start at least on Wine (I have no Windows right now so I cannot do extensive tests). You can test by switching from `jeromerobert/wine-mingw64:1.8.1` to `jeromerobert/wine-mingw64:1.9.0` in your PR `install.yml`.

I updated the Docker file according to your updated `Win32.md` except:
* `export SETUPTOOLS_USE_DISTUTILS=stdlib`
* Explicitly install `python-dateutil`

This should be added to `Win32.md`.

Anyway, thank you so much, it's now much better. Particularly:
- No need to use a patched version of `cx_Freeze` to avoid duplication of `.dll`
- `pikepdf` installation is much simpler

**Comment by kbengs** [comment]
Thanks. I'll make a PR for the update instructions and also do some testing.
